csrc/executor.cpp

```diff
@@ -1606,6 +1607,137 @@ void FusionExecutor::initializeExecutorEntry(
+        memcpy(offsets[1] + i * sizeof(int32_t), &shp, sizeof(int32_t));
+      }
+      // In rare cases we have fewer strides than shapes
+      for (size_t i = 0; i < strides.size(); ++i) {
+        const int32_t strd = static_cast<int32_t>(strides[i]);
+        memcpy(offsets[2] + i * sizeof(int32_t), &strd, sizeof(int32_t));
+      }
+    } break;
+    default:
+      NVF_CHECK(0, "Unhandled index type size");
+      break;
+  }
+}
+
+// set the arguments that we'll pass to cuLaunchKernel. This should happen
+// when we change the rank of a tensor or the number of arguments to a kernel.
+// It does not need to happen when only shapes change---use recomputeArgs for
+// that.
+void FusionExecutor::computeArgs(
+    ExecutorEntry& entry,
+    ExpressionEvaluator& expr_eval,
+    const kir::Kernel* kernel) const {
+  FUSER_PERF_SCOPE("Initial GetArgsBuffers");
+
+  const std::vector<Val*>& params = kernel->parameters();
+  entry.args.resize(params.size());
+  entry.arg_ptrs.resize(params.size());
+  const PrimDataType idx_type = kernel->indexType();
+  for (size_t p = 0; p < params.size(); ++p) {
+    entry.args[p] = getKernelArgument(expr_eval, params[p], idx_type);
+    entry.arg_ptrs[p] = entry.args[p].data();
+  }
+}
+
+// Reset the arguments that we'll pass to cuLaunchKernel. This needs to be
+// invoked on every shape change.
+void FusionExecutor::recomputeArgs(
+    ExecutorEntry& entry,
+    ExpressionEvaluator& expr_eval,
+    const kir::Kernel* kernel) const {
+  FUSER_PERF_SCOPE("Recompute GetArgsBuffers");
+  // assert(entry.init && "entry was never initialized");
+
+  const std::vector<Val*>& params = kernel->parameters();
+  const PrimDataType idx_type = kernel->indexType();
+  // assert(entry.args.size() == params.size());
+  // assert(entry.arg_ptrs.size() == params.size());
+  // assert(params.size() >= args.size());
+  for (size_t p = 0; p < params.size(); ++p) {
+    PolymorphicValue pv = expr_eval.evaluate(params[p]);
+    if (pv.is<at::Tensor>() && pv.as<at::Tensor>().is_cuda()) {
+      // GPU tensors are not passed directly: instead we pass a Tensor<type,
+      // rank, rank> struct. The pointer and dimensions are dynamic, but the
+      // types and ranks are actually static (changing the rank or the types
+      // would need to be done via a new FusionDefinition). As such, we created
+      // the Tensor<t, r, r> struct during ::computeArgs, and here we just fill
+      // in the base address, shape, and stride arrays to cover whatever new
+      // tensors we got this round.
+      TensorView* mtv = dynamic_cast<TensorView*>(params[p]);
+      const Val* mdexpr = IrBuilder::metadataExpr(mtv);
+      const PolymorphicValue& tmd = expr_eval.evaluate(mdexpr);
+      const size_t idx_type_size =
+          PrimDataType::Int == idx_type ? sizeof(int64_t) : sizeof(int32_t);
+      fill_gpu_ptr(entry, tmd, p, idx_type_size);
+    } else {
+      entry.args[p] = getKernelArgument(expr_eval, params[p], idx_type);
```

Comment:
We could maybe do better for the non-tensor arguments, too. `getKernelArgument` is what is slow, due to dynamic allocations.

Such arguments appear to be somewhat rare, though, and I think the scope of this patch is already larger than I'd like it to be.